File: daq/slc/runcontrol/src/RCCommand.cc
#include "daq/slc/runcontrol/RCCommand.h"

#include "daq/slc/runcontrol/RCState.h"

using namespace Belle2;

const RCCommand RCCommand::CONFIGURE(101, "RC_CONFIGURE");
const RCCommand RCCommand::LOAD(102, "RC_LOAD");
const RCCommand RCCommand::START(103, "RC_START");
const RCCommand RCCommand::STOP(104, "RC_STOP");
const RCCommand RCCommand::RESUME(105, "RC_RESUME");
const RCCommand RCCommand::PAUSE(106, "RC_PAUSE");
const RCCommand RCCommand::RECOVER(107, "RC_RECOVER");
const RCCommand RCCommand::ABORT(108, "RC_ABORT");
const RCCommand RCCommand::BOOT(109, "RC_BOOT");
const RCCommand RCCommand::STATUS(110, "RC_STATUS");

const RCCommand& RCCommand::operator=(const std::string& label) throw()
{
  if (NSMCommand::operator=(label) != Enum::UNKNOWN) return *this;
  else if (label == CONFIGURE.getLabel()) *this = CONFIGURE;
  else if (label == LOAD.getLabel()) *this = LOAD;
  else if (label == START.getLabel()) *this = START;
  else if (label == STOP.getLabel()) *this = STOP;
  else if (label == RECOVER.getLabel()) *this = RECOVER;
  else if (label == RESUME.getLabel()) *this = RESUME;
  else if (label == PAUSE.getLabel()) *this = PAUSE;
  else if (label == ABORT.getLabel()) *this = ABORT;
  else if (label == BOOT.getLabel()) *this = BOOT;
  else if (label == STATUS.getLabel()) *this = STATUS;
  else *this = Enum::UNKNOWN;
  return *this;
}

const RCCommand& RCCommand::operator=(int id) throw()
{
  if (NSMCommand::operator=(id) != Enum::UNKNOWN) return *this;
  else if (id == CONFIGURE.getId()) *this = CONFIGURE;
  else if (id == LOAD.getId()) *this = LOAD;
  else if (id == START.getId()) *this = START;
  else if (id == STOP.getId()) *this = STOP;
  else if (id == RECOVER.getId()) *this = RECOVER;
  else if (id == RESUME.getId()) *this = RESUME;
  else if (id == PAUSE.getId()) *this = PAUSE;
  else if (id == ABORT.getId()) *this = ABORT;
  else if (id == BOOT.getId()) *this = BOOT;
  else if (id == STATUS.getId()) *this = STATUS;
  else *this = Enum::UNKNOWN;
  return *this;
}

const RCCommand& RCCommand::operator=(const char* label) throw()
{
  if (label != NULL)  *this = std::string(label);
  else *this = Enum::UNKNOWN;
  return *this;
}

int RCCommand::isAvailable(const RCState& state) const throw()
{
  const RCCommand& cmd(*this);
  if ((cmd == LOAD || cmd == CONFIGURE) &&
      state == RCState::NOTREADY_S) {
    return SUGGESTED;
    /*  } else if ((cmd == LOAD || cmd == CONFIGURE) &&
             state == RCState::READY_S) {
    return ENABLED;
    */
  } else if (cmd == START && state == RCState::READY_S) {
    return SUGGESTED;
  } else if (cmd == STOP && (state == RCState::RUNNING_S ||
                             state == RCState::PAUSED_S)) {
    return SUGGESTED;
  } else if (cmd == PAUSE && state == RCState::RUNNING_S) {
    return ENABLED;
  } else if (cmd == RESUME && state == RCState::PAUSED_S) {
    return ENABLED;
  } else if (cmd == ABORT) {
    return ENABLED;
  } else if (cmd == RECOVER && state == RCState::ERROR_ES) {
    return ENABLED;
  } else if (cmd == BOOT && state == RCState::NOTREADY_S) {
    return ENABLED;
  } else {
    return DISABLED;
  }
}

RCState RCCommand::nextState() const throw()
{
  const RCCommand& cmd(*this);
  if (cmd == LOAD) return RCState::READY_S;
  else if (cmd == START) return RCState::RUNNING_S;
  else if (cmd == STOP) return RCState::READY_S;
  else if (cmd == RESUME) return RCState::RUNNING_S;
  else if (cmd == PAUSE) return RCState::PAUSED_S;
  else if (cmd == RECOVER) return RCState::READY_S;
  else if (cmd == ABORT) return RCState::NOTREADY_S;
  else if (cmd == BOOT) return RCState::NOTREADY_S;
  else return Enum::UNKNOWN;
}

RCState RCCommand::nextTState() const throw()
{
  const RCCommand& cmd(*this);
  if (cmd == CONFIGURE) return RCState::CONFIGURING_TS;
  else if (cmd == LOAD) return RCState::LOADING_TS;
  else if (cmd == START) return RCState::STARTING_TS;
  else if (cmd == STOP) return RCState::STOPPING_TS;
  else if (cmd == RESUME) return RCState::RUNNING_S;
  else if (cmd == PAUSE) return RCState::PAUSED_S;
  else if (cmd == RECOVER) return RCState::RECOVERING_RS;
  else if (cmd == ABORT) return RCState::ABORTING_RS;
  else if (cmd == BOOT) return RCState::BOOTING_RS;
  else return Enum::UNKNOWN;
}

File: daq/slc/extra/javalib/css/org.csstudio.belle2.daq/src/org/belle2/daq/runcontrol/RCCommand.java
package org.belle2.daq.runcontrol;

import org.belle2.daq.nsm.NSMCommand;
import org.belle2.daq.nsm.NSMState;

public class RCCommand extends NSMCommand {

    final static public RCCommand CONFIGURE = new RCCommand(101, "RC_CONFIGURE");
    final static public RCCommand LOAD = new RCCommand(102, "RC_LOAD");
    final static public RCCommand START = new RCCommand(103, "RC_START");
    final static public RCCommand STOP = new RCCommand(104, "RC_STOP");
    final static public RCCommand RESUME = new RCCommand(105, "RC_RESUMEN");
    final static public RCCommand PAUSE = new RCCommand(106, "RC_PAUSE");
    final static public RCCommand RECOVER = new RCCommand(107, "RC_RECOVER");
    final static public RCCommand ABORT = new RCCommand(108, "RC_ABORT");

    public RCCommand() {
        copy(RCCommand.UNKNOWN);
    }

    private RCCommand(int id, String label) {
        super(id, label);
    }

    public RCCommand(RCCommand command) {
        copy(command);
    }

    public RCCommand(String command) {
      copy(command);
  }
    public boolean copy(String msg) {
    	if (!msg.startsWith("RC_")) msg = "RC_"+msg;
        if (super.copy(msg)) {
            return true;
        } else if (msg.equals(CONFIGURE.getLabel())) {
            copy(CONFIGURE);
            return true;
        } else if (msg.equals(LOAD.getLabel())) {
            copy(LOAD);
            return true;
        } else if (msg.equals(START.getLabel())) {
            copy(START);
            return true;
        } else if (msg.equals(STOP.getLabel())) {
            copy(STOP);
            return true;
        } else if (msg.equals(RESUME.getLabel())) {
            copy(RESUME);
            return true;
        } else if (msg.equals(PAUSE.getLabel())) {
            copy(PAUSE);
            return true;
        } else if (msg.equals(RECOVER.getLabel())) {
            copy(RECOVER);
            return true;
        } else if (msg.equals(ABORT.getLabel())) {
            copy(ABORT);
            return true;
        }
        return false;
    }

    public int available(NSMState state) {
        if (equals(CONFIGURE) && state.equals(RCState.NOTREADY_S)) {
            return ENABLED;
        } else if (equals(LOAD) && state.equals(RCState.NOTREADY_S)) {
            return SUGGESTED;
        } else if (equals(LOAD) && state.equals(RCState.READY_S)) {
            return ENABLED;
        } else if (equals(START) && state.equals(RCState.READY_S)) {
            return SUGGESTED;
        } else if (equals(STOP)
                && (state.equals(RCState.ERROR_ES) || state.equals(RCState.FATAL_ES)
                || state.equals(RCState.RUNNING_S) || state.equals(RCState.PAUSED_S))) {
            return SUGGESTED;
        } else if (equals(PAUSE) && state.equals(RCState.RUNNING_S)) {
            return ENABLED;
        } else if (equals(RESUME) && state.equals(RCState.PAUSED_S)) {
            return SUGGESTED;
        } else if (equals(OK) || equals(ERROR) || equals(ABORT) || equals(RECOVER)
                || equals(DATAGET) || equals(DATASET) || equals(DBGET) || equals(DBSET)) {
            return ENABLED;
        } else {
            return DISABLED;
        }
    }

    public RCState nextState() {
        if (equals(LOAD)) {
            return RCState.READY_S;
        } else if (equals(START)) {
            return RCState.RUNNING_S;
        } else if (equals(STOP)) {
            return RCState.READY_S;
        } else if (equals(RESUME)) {
            return RCState.RUNNING_S;
        } else if (equals(PAUSE)) {
            return RCState.PAUSED_S;
        } else if (equals(RECOVER)) {
            return RCState.READY_S;
        } else if (equals(ABORT)) {
            return RCState.NOTREADY_S;
        } else {
            return new RCState();
        }
    }

    public RCState nextTState() {
        if (equals(LOAD)) {
            return RCState.LOADING_TS;
        } else if (equals(START)) {
            return RCState.STARTING_TS;
        } else if (equals(STOP)) {
            return RCState.STOPPING_TS;
        } else if (equals(RESUME)) {
            return RCState.RUNNING_S;
        } else if (equals(PAUSE)) {
            return RCState.PAUSED_S;
        } else if (equals(RECOVER)) {
            return RCState.RECOVERING_RS;
        } else if (equals(ABORT)) {
            return RCState.ABORTING_RS;
        } else {
            return new RCState();
        }
    }

    public boolean copy(int id) {
        if (super.copy(id)) {
            return true;
        } else if (id == CONFIGURE.getId()) {
            copy(CONFIGURE);
            return true;
        } else if (id == LOAD.getId()) {
            copy(LOAD);
            return true;
        } else if (id == START.getId()) {
            copy(START);
            return true;
        } else if (id == STOP.getId()) {
            copy(STOP);
            return true;
        } else if (id == RESUME.getId()) {
            copy(RESUME);
            return true;
        } else if (id == PAUSE.getId()) {
            copy(PAUSE);
            return true;
        } else if (id == RECOVER.getId()) {
            copy(RECOVER);
            return true;
        } else if (id == ABORT.getId()) {
            copy(ABORT);
            return true;
        }
        return false;
    }

}
